### PR TITLE
Randomshadow_interface

### PR DESCRIPTION
**Pull request by brendanator**
#1632
Improve interface for RandomShadow.
Optimize transform.

### CODE REVIEW

**Comment by brendanator** [comment]
This is a benchmark review for experiment `review_of_100_reviews_20240409`.
Run ID: `review_of_100_reviews_20240409/benchmark_2024-04-09T10-24-34_v1-16-0-95-g65fefaeef-dirty`.

This pull request was cloned from `https://github.com/albumentations-team/albumentations/pull/1637`. _(Note: the URL is not a link to avoid triggering a notification on the original pull request.)_

File: albumentations/augmentations/transforms.py [hunk 1]
```diff
@@ -841,9 +841,10 @@ class RandomShadow(ImageOnlyTransform):
     Args:
         shadow_roi: region of the image where shadows
             will appear. All values should be in range [0, 1].
-        num_shadows_lower: Lower limit for the possible number of shadows.
+        num_shadows_limit: Lower and upper limits for the possible number of shadows.
+        num_shadows_lower: Deprecated: Lower limit for the possible number of shadows.
             Should be in range [0, `num_shadows_upper`].
-        num_shadows_upper: Lower limit for the possible number of shadows.
+        num_shadows_upper: Deprecated: Lower limit for the possible number of shadows.
             Should be in range [`num_shadows_lower`, inf].
```
Comment (brendanator) [inline albumentations/augmentations/transforms.py:847]: <a href="https://sourcery.ai">Sourcery logo</a>  **issue (code_clarification):** Clarification on deprecation message for `num_shadows_upper`.

The deprecation message for `num_shadows_upper` incorrectly states 'Lower limit for the possible number of shadows.' It should indicate it's the upper limit.

File: albumentations/augmentations/transforms.py [hunk 2]
```diff
@@ -870,6 +872,17 @@ def __init__(
...
+
+        if num_shadows_lower is not None or num_shadows_upper is not None:
+            warn(
+                "`num_shadows_lower` and `num_shadows_upper` are deprecated. "
+                "Use `num_shadows_limit` as tuple (num_shadows_lower, num_shadows_upper) instead.",
+                DeprecationWarning,
+                stacklevel=2,
+            )
+        num_shadows_lower = num_shadows_lower or num_shadows_limit[0]
+        num_shadows_upper = num_shadows_upper or num_shadows_limit[1]
```
Comment (brendanator) [inline albumentations/augmentations/transforms.py:884]: <a href="https://sourcery.ai">Sourcery logo</a>  **suggestion (edge_case_not_handled):** Consider edge case where both deprecated parameters are used.

If both `num_shadows_lower` and `num_shadows_upper` are provided, this logic will override them with `num_shadows_limit`. It might be worth handling or warning about this specific case.

File: tests/test_transforms.py [hunk 3]
```diff
@@ -1416,7 +1417,7 @@ def test_deprecation_warnings(size, width, height, expected_warning):
     warnings.resetwarnings()
 
 
-def test_randomgridshuffle():
+def test_randomgridshuffle() -> None:
```
Comment (brendanator) [inline tests/test_transforms.py:1420]: <a href="https://sourcery.ai">Sourcery logo</a>  **suggestion (testing):** Consider adding a test case for `RandomShadow` transform.

It appears that the changes to `RandomShadow` are significant, particularly around the handling of `num_shadows_limit`, `num_shadows_lower`, and `num_shadows_upper`. A dedicated test case that verifies the correct behavior of shadows being added to the image, considering these parameters, would ensure the functionality works as expected.

```suggestion
@pytest.mark.parametrize("num_shadows, expected_num_shadows", [
    ((1, 2), random.randint(1, 2)),
    ((2, 3), random.randint(2, 3)),
])
def test_random_shadow_num_shadows(
    num_shadows: Tuple[int, int], expected_num_shadows: int
) -> None:
    shadow = A.RandomShadow(num_shadows_limit=num_shadows)
    assert shadow.num_shadows == expected_num_shadows
```

File: tests/test_transforms.py [hunk 4]
```diff
@@ -1428,3 +1429,37 @@ def test_randomgridshuffle():
...
+                               num_shadows_upper=num_shadows_upper)
+        else:
+            A.RandomShadow(num_shadows_limit=num_shadows_limit, num_shadows_lower=num_shadows_lower,
+                           num_shadows_upper=num_shadows_upper)
+        if expected_warning is DeprecationWarning:
+            assert len(w) == 1
+            assert issubclass(w[-1].category, expected_warning)
+        else:
+            assert not w
+    warnings.resetwarnings()
```
Comment (brendanator) [inline tests/test_transforms.py:1465]: <a href="https://sourcery.ai">Sourcery logo</a>  **suggestion (testing):** Add test cases to verify the actual transformation effect of `RandomShadow`.

While the current tests validate parameter handling and deprecation warnings, there are no tests checking the visual outcome of the `RandomShadow` transformation. Adding such tests could involve checking if the shadow is correctly applied within the specified `shadow_roi` and if the number of shadows matches `num_shadows_limit`.

Comment (brendanator) [inline tests/test_transforms.py:1465]: <a href="https://sourcery.ai">Sourcery logo</a>  **suggestion (testing):** Ensure edge cases for `num_shadows_limit` are tested.

The current parameterized test covers a range of scenarios for `num_shadows_limit`, including valid ranges and deprecation warnings. However, it would be beneficial to include edge cases where `num_shadows_limit` is at its minimum and maximum valid values, as well as cases where the limits are equal, to ensure these scenarios are handled correctly.

File: albumentations/augmentations/functional.py [hunk 5]
```diff
@@ -740,14 +740,14 @@ def add_sun_flare(
 
 @ensure_contiguous
 @preserve_shape
-def add_shadow(img: np.ndarray, vertices_list: List[List[Tuple[int, int]]]) -> np.ndarray:
```
Comment (brendanator) [inline albumentations/augmentations/functional.py:743]: <a href="https://sourcery.ai">Sourcery logo</a>  **suggestion (docstrings):** Please update the docstring for function: `add_shadow`

Reason for update: The type of `vertices_list` argument has changed from a list of lists of tuples to a list of numpy arrays.

Suggested new docstring:
```
"""Add shadows to the image.

From https://github.com/UjjwalSaxena/Automold--Road-Augmentation-Library

Args:
    img (numpy.ndarray):
    vertices_list (list[numpy.ndarray]):

Returns:
    numpy.ndarray:
"""
```

**Review by brendanator — COMMENTED** [review]
<a href="https://sourcery.ai">Sourcery logo</a>  Sourcery Review:

Hey @brendanator - I've reviewed your changes and they look great!

<details>
<summary>Here's what I looked at during the review</summary>

- 🟡 **General issues**: 2 issues found
- 🟢 **Security**: all looks good
- 🟡 **Testing**: 3 issues found
- 🟢 **Complexity**: all looks good
- 🟡 **Docstrings**: 1 issue found
</details>

***

[LangSmith trace](https://smith.langchain.com/o/5c8bea2b-6c6c-4221-bd9b-0c11556933a8/projects/p/2dbf7697-3bb3-4f63-b184-3a8ebf7ac23c/r/9e54c321-a191-4e4d-82ce-e2355970f71f?poll=true)

<sub>
Help me be more useful! Please click 👍 or 👎 on each comment to tell me if it was helpful.
</sub>
